src/corehost/common/trace.h

```diff
@@ -31,6 +33,39 @@ namespace trace
 
     // Returns the currently set callback for error writing
     error_writer_fn get_error_writer();
+
+    void trace(const pal::char_t* format, ...);
+    void trace_error(const pal::char_t* format, ...);
+
+    void println(const pal::char_t* format, ...);
+    void println();
+    void flush();
 };
 
+// Tracing is performed with a macro so that arguments to trace::trace() and
+// trace::trace_error() can be evaluated lazily depending on the current
+// verbosity level.
+#define TRACE_LAZY(Verbosity, Args...)                              \
```

Comment:
I would strongly avoid this pattern. It adds a lot to code bloat and isn't debuggable. Argument evaluation isn't a measurable cost here if all are pointer sized.

Reply:
In this entire change there are no costly argument evaluations. Logging should never need to evaluate complex arguments so I am really confused as to this concern.

Reply:
The suggestion came from profiling. During normal console app startup we do cca 700 `version_t.as_str()` which are pretty much all due to tracing. The method turns 4 numbers into a string a concatenates them - which is not exactly cheap.
I think that tracing might do semi-expensive things, that is turn structures to strings - so that it's humanly readable - this typically involves allocations and so on.
If you have an idea how to avoid it... I'm all for it. I looked on the internet and the recommendations are generally either macros (similar to this PR) or lambdas. There might be a solution based on C++ templates, but that seems to be rather tricky.

One question - why do you think this introduces a code bloat? All the macro does is add a simple if, which the compiler will very likely optimize to a simple memory read, comparison and branch.

Reply:
Oh, I must have missed the `.as_str()`. I only saw a series of `.c_str()`. If that is the only scenario I would look for other solutions - templates specifically or in a pinch operator overloads for the classes in question.

Lambdas fine, but can have serious performance impact when used with capture - non-capture lambdas are generally fine.

The code bloat is related to black box that is a macro and inevitable additional code that will be inserted. There are patterns that are return value checks (i.e. `RETURN_IF_FAILED()`) which tend to be clear, but logging eventually turns into a grab bag of other features for additional logging and state recording during a crash or perhaps evaluation of asserts to aid in some other scenario. Basically logging and macros breed code bloat because "it's logging and we need more details".

This is based purely on my experience, but I think CoreCLR does a good job at demonstrating my concerns independent of my experience.

Reply:
This actually reduces some of the code bloat.

The outermost `if` is evaluated at compile time (both sides in the comparisons are compile-time constants).  It could be a `if constexpr`, but we're not using C++17 yet.

In the non-error cases, calls to `current_verbosity()` should be hoisted -- it's marked as a pure function, so it should be just a single function call if the compiler decides not to inline it -- and that should be pretty much it: the rest will be either just integer comparisons, or, most likely, just one integer comparison for all the calls in the same verbosity within a function (due to constant propagation/folding).

Also, instead of having multiple versions of tracing functions, there's essentially just two: one for the error path, and other for all the others.  Having less code around is a good thing.

The error path is different, and it should be rare; that's why I put it into a lambda that's declared and called immediately (and marked as "noinline" where supported so it's far away from the happy path as far as things like branch predictors, prefetchers, uop caches and this sort of thing), capturing only by reference.  Most of the cases I saw were capturing a single `std::string.c_str()` or an integer type, so it should be fine.  I don't care much about the performance of the error path, as long as it doesn't interfere with the happy path -- that's why this idiom was used.

I've investigated other things, including using variadic templates.  I contemplated writing my own printf-like function that takes C++ types directly, and this remains an option here, considering that all tracing calls use very basic conversion specifiers (`%s`, `%d`, and `%x` -- with the occasional `%04x`). This could support `std::string` directly (removing the need to call `c_str()`, which makes the call sites cleaner), or arbitrary types with `operator std::string()` that could be implemented to perform expensive to-string operations only when necessary.  I decided to give the macro version a try first because it didn't require any changes to any of the calling sites that couldn't be done with an invocation of `sed` in the command line; but if people prefer this other method, I can try to come up with something.

Reply:
>This actually reduces some of the code bloat.

My comment wasn't to be taken about present state, but precedent. Agreed that there are options to reduce using `constexpr`, but those aren't actually real because we don't want compile time optimization to exclude code because we want to enable logging on release builds as well so `constexpr` isn't something we can leverage - at least to my understanding.

>should be hoisted -- it's marked as a pure function, so it should be just a single function call if the compiler decides not to inline it 

Fully appreciate what the compiler can do, but again that isn't my concern. It is about precedence and what is needed. If logging is actually take more than 1 % when it is off then we are doing something wrong in general and playing optimization games tends to create fragile code paths that can easily become degenerate.

This leads into another principle - can most people look at this and understand how to maintain it? Comments on instructions caches tend to indicate code that relies on too many assumptions or deep knowledge about something that defeats the purpose of higher level languages. In this case we are talking about logging so if this is where we feel we need to hand optimize to that level then there should be literally no other hot spots in the code.

In the GC, JIT, type load, all of these comments and optimizations seem spot on. Not for logging which can be addressed with some basic templates that rely entirely on opaque compiler optimizations and not require any comments or descriptions about what the compiler can/should/may do.

> I decided to give the macro version a try first because it didn't require any changes to any of the calling sites that couldn't be done with an invocation of `sed` in the command line; but if people prefer this other method, I can try to come up with something.

Based on what I am seeing, I think the function should be convertible to a template without much code change, but of course that depends on how it is done. I would prefer to try something else if possible.